### DISCUSSION TITLE
Getting `add_raster` to work in a remote host in vscode: port forwarding.

### DISCUSSION THREAD

**Discussion by lopezvoliver** [opengeos/leafmap#677] [Show and tell]
I usually work with [vscode](https://code.visualstudio.com/) and [connect to a remote server](https://code.visualstudio.com/docs/remote/ssh). 

I had some initial trouble getting `add_raster` to work for me in this setup, so I will share here what ended up working for me. 

In a nutshell, `add_raster` leverages  [localtileserver](https://github.com/banesullivan/localtileserver), and one of the keyword arguments that can get passed on to it is the `port` number:

>    port : int
>        The port on your host machine to use for the tile server (if creating
>       a tileserver. This is ignored if a file path is given). This defaults
>       to getting an available port.

In my experience, vscode typically does a good job of automatically forwarding ports opened by your applications. However, in my case this wasn't working and my raster didn't get displayed on the map. You can manually forward the port, but if you didn't specify a port number to use, how do you figure out which one was opened?

One solution is to print the layer url:

```
print(m.layers[1].url)
```

where `m` is your `leafmap.Map` and `m.layers[1]` is a layer you added using `add_raster`. This will print something like:

```
http://127.0.0.1:34389/api/tiles/{z}/{x}/{y}.png?&filename=%2Frhome%2Flopezom%2Flandsat7.tif&indexes.0=1&indexes.1=2&indexes.2=3&nodata=-1](http://127.0.0.1:34389/api/tiles/%7Bz%7D/%7Bx%7D/%7By%7D.png?&filename=%2Frhome%2Flopezom%2Flandsat7.tif&indexes.0=1&indexes.1=2&indexes.2=3&nodata=-1
```

in this case the port number is 34389. Then you can manually forward it using vscode's `Forward a Port` command:

![image](https://github.com/opengeos/leafmap/assets/14804652/f18402b5-e160-4893-8fd5-57c1d4e163ba)

![image](https://github.com/opengeos/leafmap/assets/14804652/95fe1f0e-0872-4047-8f11-da9b8cddf680)

After this, try moving the map or zooming in/out and hopefully you will now start to see some tiles show up in the map. 

A simpler solution is to first manually forward a port (e.g. 7778) and specify it in `add_raster(..., port=port)`.

**Comment by giswqs** [comment]
Thank you for sharing the solution! That's very help. I am going to pin the discussion at the top so that others can learn from this.

**Comment by lopezvoliver** [comment]
Another solution is to use the command `Developer: Toggle Developer Tools`

![image](https://github.com/opengeos/leafmap/assets/14804652/6911e818-d12c-4359-8648-69a5b25a1fc0)

which will open a window like this:

![image](https://github.com/opengeos/leafmap/assets/14804652/3ec4d8ef-4bc0-430c-a747-a701c768e066)

In this example you can identify that the port is 39421. Then you can forward it and you will see the tiles on the map.
